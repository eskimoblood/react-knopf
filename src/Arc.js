import React from 'react';

function Arc(props) {
  const startAngle = (props.angleOffset || 0) - 90;
  const outerRadius = props.outerRadius || props.center;
  const innerRadius = props.innerRadius || props.center - props.width;
  const startAngleDegree = toRad(startAngle);
  const endAngleDegree = toRad(startAngle + props.angle);

  const p1 = pointOnCircle(props.center, outerRadius, endAngleDegree);
  const p2 = pointOnCircle(props.center, outerRadius, startAngleDegree);
  const p3 = pointOnCircle(props.center, innerRadius, startAngleDegree);
  const p4 = pointOnCircle(props.center, innerRadius, endAngleDegree);
  const largeArcFlag = ( props.angle < 180 ? 0 : 1);

  const moveToP1 = moveTo(p1);
  const arcToP2 = arcTo(p2, outerRadius, largeArcFlag, 0);
  const lineToP3 = lineTo(p3);
  const arcToP4 = arcTo(p4, innerRadius, largeArcFlag, 1);
  const lineToP1 = lineTo(p1);

  return <path d={`${moveToP1} ${arcToP2} ${lineToP3} ${arcToP4} ${lineToP1}`}/>;
}

function moveTo(p) {
  return `M${p.x},${p.y}`;
}

function arcTo(p, radius, largeArcFlag, sweepFlag) {
  return `A${radius},${radius} 0 ${largeArcFlag} ${sweepFlag} ${p.x},${p.y}`;
}

function lineTo(p) {
  return `L${p.x},${p.y}`
}

function toRad(angle) {
  return Math.PI * angle / 180;
}

function pointOnCircle(center, radius, angle) {
  return {
    x: center + radius * Math.cos(angle),
    y: center + radius * Math.sin(angle)
  };
}

export default Arc;