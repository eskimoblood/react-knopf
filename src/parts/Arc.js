import React, { PropTypes } from 'react';
import { toRad, arcTo, lineTo, moveTo, pointOnCircle } from '../utils/arcUtils';

function Arc(props) {
  const startAngle = props.angleOffset - 90;
  const outerRadius = props.outerRadius || props.center;
  const innerRadius = props.innerRadius || props.center - props.knobSize / 15;
  const startAngleDegree = toRad(startAngle);
  const endAngleDegree = toRad(props.angle - 90);


  const p1 = pointOnCircle(props.center, outerRadius, endAngleDegree);
  const p2 = pointOnCircle(props.center, outerRadius, startAngleDegree);
  const p3 = pointOnCircle(props.center, innerRadius, startAngleDegree);
  const p4 = pointOnCircle(props.center, innerRadius, endAngleDegree);
  const largeArcFlag = (props.angle - props.angleOffset) < 180 ? 0 : 1;

  const moveToP1 = moveTo(p1);
  const arcToP2 = arcTo(p2, outerRadius, largeArcFlag, 0);
  const lineToP3 = lineTo(p3);
  const arcToP4 = arcTo(p4, innerRadius, largeArcFlag, 1);
  const lineToP1 = lineTo(p1);

  var d = `${moveToP1} ${arcToP2} ${lineToP3} ${arcToP4} ${lineToP1}`;

  return <path d={d} fill={props.fill} filter={props.filter} />;
}

Arc.propTypes = {
  outerRadius: PropTypes.number,
  innerRadius: PropTypes.number,
  center: PropTypes.number,
  angle: PropTypes.number,
  angleOffset: PropTypes.number,
  knobSize: PropTypes.number
};

export default Arc;
