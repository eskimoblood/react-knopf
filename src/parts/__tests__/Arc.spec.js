import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import Arc from '../Arc';

describe('Arc', () => {
  describe('renders correctly', () => {

    it('for small angels', () => {
      const arc = shallow(
        <Arc
          angleOffset={0}
          outerRadius={100}
          innerRadius={90}
          knobSize={100}
          center={0}
          angle={90}
          fill="lime"
          filter="filter1"
        />);
      expect(shallowToJson(arc)).toMatchSnapshot();
    });

    it('for large angels', () => {
      const arc = shallow(
        <Arc
          angleOffset={0}
          outerRadius={100}
          innerRadius={90}
          knobSize={100}
          center={0}
          angle={260}
          fill="lime"
          filter="filter1"
        />);
      expect(shallowToJson(arc)).toMatchSnapshot();
    });
  });
});



