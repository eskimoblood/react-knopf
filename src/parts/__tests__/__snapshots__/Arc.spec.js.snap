exports[`Arc renders correctly for large angels 1`] = `
<path
  d="M-98.4807753012208,17.364817766693026 A100,100 0 1 0 6.123233995736766e-15,-100 L5.5109105961630896e-15,-90 A90,90 0 1 1 -88.63269777109872,15.628335990023725 L-98.4807753012208,17.364817766693026"
  fill="lime"
  filter="filter1" />
`;

exports[`Arc renders correctly for small angels 1`] = `
<path
  d="M100,0 A100,100 0 0 0 6.123233995736766e-15,-100 L5.5109105961630896e-15,-90 A90,90 0 0 1 90,0 L100,0"
  fill="lime"
  filter="filter1" />
`;
